dds/src/implementation/dds_impl/builtin_publisher.rs

```diff
@@ -1,51 +1,64 @@
 use crate::{
     implementation::{
+        data_representation_builtin_endpoints::{
+            discovered_reader_data::{DiscoveredReaderData, DCPS_SUBSCRIPTION},
+            discovered_topic_data::{DiscoveredTopicData, DCPS_TOPIC},
+            discovered_writer_data::{DiscoveredWriterData, DCPS_PUBLICATION},
+        },
         rtps::{
+            endpoint::RtpsEndpoint,
             group::RtpsGroup,
             messages::submessages::AckNackSubmessage,
-            types::{EntityId, EntityKey, Guid, GuidPrefix, Locator, BUILT_IN_WRITER_GROUP},
+            stateful_writer::{
+                RtpsStatefulWriter, DEFAULT_HEARTBEAT_PERIOD, DEFAULT_NACK_RESPONSE_DELAY,
+                DEFAULT_NACK_SUPPRESSION_DURATION,
+            },
+            types::{
+                EntityId, EntityKey, Guid, GuidPrefix, Locator, TopicKind, BUILT_IN_WRITER_GROUP,
+            },
+            writer::RtpsWriter,
         },
-        utils::{
-            condvar::DdsCondvar,
-            shared_object::{DdsRwLock, DdsShared},
+        utils::shared_object::{DdsRwLock, DdsShared},
+    },
+    infrastructure::{
+        error::DdsResult,
+        qos::{DataWriterQos, PublisherQos},
+        qos_policy::{
+            DurabilityQosPolicy, DurabilityQosPolicyKind, HistoryQosPolicy, HistoryQosPolicyKind,
+            ReliabilityQosPolicy, ReliabilityQosPolicyKind,
         },
+        status::NO_STATUS,
+        time::{DurationKind, DURATION_ZERO},
     },
-    infrastructure::{error::DdsResult, qos::PublisherQos},
+    topic_definition::type_support::DdsType,
 };
 
 use super::{
-    builtin_stateful_writer::BuiltinStatefulWriter,
     builtin_stateless_writer::BuiltinStatelessWriter,
     domain_participant_impl::{
         ENTITYID_SEDP_BUILTIN_PUBLICATIONS_ANNOUNCER,
         ENTITYID_SEDP_BUILTIN_SUBSCRIPTIONS_ANNOUNCER, ENTITYID_SEDP_BUILTIN_TOPICS_ANNOUNCER,
         ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER,
     },
     message_receiver::{MessageReceiver, PublisherMessageReceiver},
-    topic_impl::TopicImpl,
+    user_defined_data_writer::UserDefinedDataWriter,
 };
 
 pub struct BuiltinPublisher {
     _qos: PublisherQos,
     _rtps_group: RtpsGroup,
     spdp_builtin_participant_writer: DdsShared<BuiltinStatelessWriter>,
-    sedp_builtin_topics_writer: DdsShared<BuiltinStatefulWriter>,
-    sedp_builtin_publications_writer: DdsShared<BuiltinStatefulWriter>,
-    sedp_builtin_subscriptions_writer: DdsShared<BuiltinStatefulWriter>,
+    sedp_builtin_topics_writer: DdsShared<UserDefinedDataWriter<RtpsStatefulWriter>>,

```

Comment:
That the builtin is of type UserDefined is a little confusing

Reply:
I was planning to rename it later but I have done it now instead :)